### PR TITLE
Loop pedal

### LINKED ISSUE #17: Loop pedal

### PR DESCRIPTION
**Pull request by peterstuart**
As discussed, this is the most basic version with just 3 states: off, recording, playing.

Not done (I'll file issues for these):

- Overdubbing
- Key events are always handled at the beginning of the next audio frame, but depending on the frame size, that may not be accurate enough. (I'm not sure how to correlate the timestamp on MIDI messages to the current time.)

Closes #17.

### CODE REVIEW

File: src/audio_unit/delay.rs [hunk 1]
```diff
@@ -48,11 +48,6 @@ impl Delay {
         ))
     }
 
-    fn delay_num_samples(stream_config: &StreamConfig, delay_ms: DelayMs) -> usize {
-        let delay_num_frames = (delay_ms as f32 / 1_000.0) * stream_config.sample_rate.0 as f32;
-        delay_num_frames as usize * stream_config.channels as usize
-    }
-
```
Comment (peterstuart) [inline src/audio_unit/delay.rs:55]: Moved this to `util` since it is generally useful.

File: src/audio_unit/looper.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    fn process_samples(&mut self, input: &[f32], output: &mut [f32]) {
+        match self.state {
+            Off => {
+                util::zero_slice(output);
+            }
+            Recording { position } => {
+                let next_position = position + input.len();
+
+                if next_position <= self.buffer.len() {
```
Comment (peterstuart) [inline src/audio_unit/looper.rs:74]: If this entire frame can be handled as one thing (because it doesn't exceed the max size we have for a recording)

File: src/audio_unit/looper.rs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+                            position: 0,
+                            total: next_position,
+                        };
+                    } else {
+                        self.state = Recording {
+                            position: next_position,
+                        };
+                    }
+                } else {
+                    self.process_samples_wrap_around(position, self.buffer.len(), input, output);
```
Comment (peterstuart) [inline src/audio_unit/looper.rs:90]: If not, break it into two parts and handle them separately.

File: src/audio_unit/looper.rs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+                    self.state = Playing {
+                        position: next_position % total,
+                        total,
+                    }
+                } else {
+                    self.process_samples_wrap_around(position, total, input, output);
+                }
+            }
+        };
+    }
```
Comment (peterstuart) [inline src/audio_unit/looper.rs:108]: This function is a little hard to follow (and was hard to write). Any ideas on simplifying this?

File: src/effect/looper.rs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    split: audio_unit::Split,
+    messages: Sender<Message>,
+}
+
+impl Looper {
+    pub fn new(config: LooperConfig, stream_config: &StreamConfig) -> Result<Self> {
+        let (looper, messages) = audio_unit::Looper::new(stream_config, config.max_ms);
+        let transparent = audio_unit::Transparent::new();
+
+        let split = audio_unit::Split::new(vec![looper.boxed(), transparent.boxed()])?;
```
Comment (peterstuart) [inline src/effect/looper.rs:23]: Like you mentioned before, it's kind of weird that every effect has to do this, but I didn't want to deal with that in this PR.

Comment (helixbass) [reply src/effect/looper.rs:23]: And likely lots of types of effects wouldn't actually pass through the clean signal (whether they have a "wet/dry" concept or just always mess with the incoming signal)

File: src/audio_unit/looper.rs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+#[derive(Debug)]
+pub enum Message {
+    Toggle,
+}
+
+#[derive(Debug)]
+enum State {
+    Off,
+    Recording { position: usize },
```
Comment (helixbass) [inline src/audio_unit/looper.rs:14]: What really is a `usize`? Is it always an "index" into eg an array and if so why is that a special type? I feel like I've read that it's equivalent to a pointer type (ie it can fit a memory address), but indexes aren't pointers?

File: src/audio_unit/looper.rs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+            Off => Recording { position: 0 },
+            Recording { position } => Playing {
+                position: 0,
+                total: position,
+            },
+            Playing {
+                position: _,
+                total: _,
+            } => {
+                util::zero_slice(&mut self.buffer);
```
Comment (helixbass) [inline src/audio_unit/looper.rs:58]: Trying to improve my mental model of the buffers a bit - why do you need to reset the buffer to zero's when turning off the looper?

File: src/audio_unit/looper.rs [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+                }
+            }
+            Playing { position, total } => {
+                let next_position = position + output.len();
+
+                if next_position <= total {
+                    output.copy_from_slice(&self.buffer[position..next_position]);
+
+                    self.state = Playing {
+                        position: next_position % total,
```
Comment (helixbass) [inline src/audio_unit/looper.rs:100]: So the `%` here is just handling the `next_position == total` -> `0` case?

File: src/audio_unit/looper.rs (hunk 4 above)
Comment (helixbass) [reply src/audio_unit/looper.rs:108]: I think it reads pretty well

File: src/audio_unit/looper.rs [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+        input: &[f32],
+        output: &mut [f32],
+    ) {
+        let split_position = total - position;
+
+        // play/record to end of buffer
+        self.process_samples(&input[0..split_position], &mut output[0..split_position]);
+
+        // ... then wrap around
+        self.process_samples(
```
Comment (helixbass) [inline src/audio_unit/looper.rs:123]: Ok so because `process_samples()` automatically switches to `Playing` once it (exactly) hits the end of its buffer size, this will now be in `Playing` mode (so result in playback of the first however many samples in the buffer) regardless of whether we entered `process_samples_wrap_around()` in `Recording` or `Playing` mode?

**Review by helixbass — APPROVED** [review]
@peterstuart very cool to see this working. My instinct still goes to messing with the MIDI clock from here and eg automatically setting the loop duration to the nearest bar line or quarter note

File: src/audio_unit/looper.rs (hunk 8 above)
Comment (peterstuart) [reply src/audio_unit/looper.rs:100]: Mmm, yes, I guess so. This is probably leftover from an earlier version before `process_samples_wrap_around` existed. I'll change it to be more clear.

File: src/effect/looper.rs (hunk 5 above)
Comment (peterstuart) [reply src/effect/looper.rs:23]: Hmm, in that case does it actually make sense to leave it up the effect then? Maybe with a helper to remove some of the boilerplate?

File: src/audio_unit/looper.rs (hunk 7 above)
Comment (peterstuart) [reply src/audio_unit/looper.rs:58]: I guess you don't, since to get back into a playback state, you'd have to go through `Recording` which would overwrite everything anyway. I'll update it (after verifying that's true).

File: src/audio_unit/looper.rs (hunk 6 above)
Comment (helixbass) [reply src/audio_unit/looper.rs:14]: I still don't get why an index == a pointer, one is a small integer (that doesn't eg have to get directly added to a pointer address in pointer arithmetic - I guess indexes would get _multiplied_ by a pointer size in pointer arithmetic, but does that imply it needs to be the same type as the thing it's being multiplied by?) and the other is a memory address

File: src/effect/looper.rs (hunk 5 above)
Comment (helixbass) [reply src/effect/looper.rs:23]: > Hmm, in that case does it actually make sense to leave it up the effect then? Maybe with a helper to remove some of the boilerplate?

Ya that sounds like the right approach ("opt-in")

File: src/audio_unit/looper.rs (hunk 6 above)
Comment (peterstuart) [reply src/audio_unit/looper.rs:14]: I think it would need to be casted/converted to the same type as the pointer to do math with it (at least to do that math efficiently).

It does feel a little leaky in a sense, but I'm not sure what the alternative would be. I think the key is that "one is a small integer" is not true.

On a 32-bit CPU, the maximum length of a `Vec` is (handwave, handwave) 2^32 (a `Vec` of `bool`s, which consumes the entire addressable memory), so the index type needs to be >= `u32`.

On a 64-bit CPU, the maximum length of a `Vec` is 2^64, so the index type needs to be >= `u64`.

So the options are:

- `u32` on all architectures:
  - good on 32-bit
  - arbitrarily restricts the length of `Vec`s on 64-bit
- `u64` on all architectures:
  - works on 32-bit (I think), but with a significant performance penality when doing math, since it's not a native type to the CPU. also, it's impossible to use the top 32-bits of the index because your program would always crash when trying to allocate an array longer than 2^32
  - good on 64-bit
- arbitrary precision integer type
  - worst of both previous choices: *way* more expensive than native types on all architectures
  - no gains (the amount of addressable memory doesn't change)
- `usize`
  - allows full usage of the addressable memory on both architectures
  - efficient on both architectures since it will always be a "native" type

Comment (peterstuart) [reply src/audio_unit/looper.rs:14]: arbitrary precision: "no gains" other than consistency, I guess

File: src/audio_unit/looper.rs (hunk 9 above)
Comment (peterstuart) [reply src/audio_unit/looper.rs:123]: Right
